decorate.go

```diff
@@ -0,0 +1,250 @@
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package dig
+
+import (
+	"fmt"
+	"reflect"
+
+	"go.uber.org/dig/internal/digreflect"
+	"go.uber.org/dig/internal/dot"
+)
+
+type decorator interface {
+	Call(c containerStore) error
+	ID() dot.CtorID
+}
+
+type decoratorNode struct {
+	dcor  interface{}
+	dtype reflect.Type
+
+	id dot.CtorID
+
+	// Location where this function was defined.
+	location *digreflect.Func
+
+	// Whether the decorator owned by this node was already called.
+	called bool
+
+	// Parameters of the decorator
+	params paramList
+
+	// Results of the decorator
+	results resultList
+
+	// order of this node in each Scopes' graphHolders.
+	orders map[*Scope]int
+
+	// scope this node was originally provided to.
+	s *Scope
+}
+
+func newDecoratorNode(dcor interface{}, s *Scope) (*decoratorNode, error) {
+	dval := reflect.ValueOf(dcor)
+	dtype := dval.Type()
+	dptr := dval.Pointer()
+
+	// Create parameter / result list.
+	pl, err := newParamList(dtype, s)
+	if err != nil {
+		return nil, err
+	}
+
+	rl, err := newResultList(dtype, resultOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	n := &decoratorNode{
+		dcor:     dcor,
+		dtype:    dtype,
+		id:       dot.CtorID(dptr),
+		location: digreflect.InspectFunc(dcor),
+		orders:   make(map[*Scope]int),
+		params:   pl,
+		results:  rl,
+		s:        s,
+	}
+	return n, nil
+}
+
+func (n *decoratorNode) Call(s containerStore) error {
+	if n.called {
+		return nil
+	}
+
+	if err := shallowCheckDependencies(s, n.params); err != nil {
+		return errMissingDependencies{
+			Func:   n.location,
+			Reason: err,
+		}
+	}
+
+	args, err := n.params.BuildList(n.s, true)
+	if err != nil {
+		return errArgumentsFailed{
+			Func:   n.location,
+			Reason: err,
+		}
+	}
+
+	results := reflect.ValueOf(n.dcor).Call(args)
+	if err := n.results.ExtractList(n.s, true, results); err != nil {
+		return err
+	}
+	n.called = true
```

Comment:
I think it's technically more correct to set it here once we confirmed there is no error. 

For example, you could imagine doing something like trying to decorate with a function whose dependencies haven't been met - in such case, line 94 or 101 will fail. But we shouldn't mark this decorator as "called" because it'll simply return without attempting to call it again. 

Basically marking called as true earlier than here would mean the following scenario would break (when it's not supposed to):
1. Provide only part of the dependency of a decorator function f. 
2. Try to decorate using f. 
3. Call something that uses f produces. (and run into an error)
4. Provide the remainder of the dependency of f.
5. Call that something again from step 3. 
6. We should expect that second call to succeed.